### PR TITLE
OSDOCS#6020: cert-manager: Customizing API fields

### PR DESCRIPTION
**Pull request by xenolinux**
Version(s): 4.12+

Issue: https://issues.redhat.com/browse/OSDOCS-6020

Link to docs preview: https://59726--docspreview.netlify.app/openshift-enterprise/latest/security/cert_manager_operator/cert-manager-customizing-api-fields.html

QE review:
- [x] QE has approved this change.

SME review:
- [x] SME has approved this change.

Additional information:

### CODE REVIEW

File: modules/cert-manager-override-environment-variables.adoc
Comment (swghosh) [inline modules/cert-manager-override-environment-variables.adoc:54]: Can we use a yaml output of the pods instead? So users could verify if their newly added args/env have come up in the pod spec.
```sh
$ oc get pods -n cert-manager -o yaml
```
should be good, IMO.

File: _topic_maps/_topic_map.yml
Comment (swghosh) [inline _topic_maps/_topic_map.yml:994]: Please re-title this to "Customizing cert-manager using operator API fields"

File: modules/cert-manager-override-arguments.adoc
Comment (swghosh) [inline modules/cert-manager-override-arguments.adoc:7]: Probably, this one could also be re-phrased a bit. How about:
`Customizing cert-manager by overriding arguments from operator API`

File: modules/cert-manager-override-environment-variables.adoc
Comment (swghosh) [inline modules/cert-manager-override-environment-variables.adoc:7]: Similar here, how about?
`Customizing cert-manager by overriding environment variables from operator API`

File: modules/cert-manager-override-arguments.adoc
Comment (swghosh) [inline modules/cert-manager-override-arguments.adoc:59]: Same as the other one,
```
$ oc get pods -n cert-manager -o yaml
```

**Review by swghosh — CHANGES_REQUESTED** [review]
In good shape, just a few minor nits. Thanks!

File: modules/cert-manager-override-arguments.adoc
Comment (xenolinux) [reply modules/cert-manager-override-arguments.adoc]: Updated '--v=<vlevel>'  as a --v=<verbosity_level>

**Comment by xenolinux** [comment]
@swghosh Can you please re-review this PR? I addressed the suggestions

File: modules/cert-manager-override-arguments.adoc
Comment (swghosh) [reply modules/cert-manager-override-arguments.adoc:58]: I checked your recent change again wrt, but am leaning more towards the yaml output tbh, the rationale here being that in the verification step users would want to check if the args/env they set have been applied on the actual pod(s) or not.
I understand that the output is way too much, but can you snip it for relevant parts and still put it, i.e. eg.

```log
$ oc get pods -n cert-manager -o yaml
---
    name: cert-manager-6d4b5d4c97-kldwl
    namespace: cert-manager
---
  spec:
    containers:
    - args:
      - --v=2
      - --cluster-resource-namespace=$(POD_NAMESPACE)
      - --leader-election-namespace=kube-system
 ---
    name: cert-manager-cainjector-866c4fd758-ltxxj
    namespace: cert-manager
---
  spec:
    containers:
    - args:
      - --v=2
      - --leader-election-namespace=kube-system
---
    name: cert-manager-webhook-6d48f88495-c88gd
    namespace: cert-manager
---
  spec:
    containers:
    - args:
      - --v=2
      - --secure-port=10250
      - --dynamic-serving-ca-secret-namespace=$(POD_NAMESPACE)
      - --dynamic-serving-ca-secret-name=cert-manager-webhook-ca
      - --dynamic-serving-dns-names=cert-manager-webhook,cert-manager-webhook.$(POD_NAMESPACE),cert-manager-webhook.$(POD_NAMESPACE).svc
---
```

File: modules/cert-manager-override-environment-variables.adoc
Comment (swghosh) [reply modules/cert-manager-override-environment-variables.adoc:51]: Similar to my the other comment, except for this one you show the environment variables only.
```
$ oc get pods -n cert-manager -o yaml
---
    name: cert-manager-6d48f88495-c88gd
    namespace: cert-manager
---
env:
      - name: POD_NAMESPACE
        valueFrom:
          fieldRef:
            apiVersion: v1
            fieldPath: metadata.namespace
      - name: HTTP_PROXY
---
```

File: modules/cert-manager-override-environment-variables.adoc
Comment (swghosh) [inline modules/cert-manager-override-environment-variables.adoc:36]: This should be Env instead of Args here.
```suggestion
    overrideEnv:
```

**Review by swghosh — CHANGES_REQUESTED** [review]

File: modules/cert-manager-override-environment-variables.adoc
Comment (swghosh) [inline modules/cert-manager-override-environment-variables.adoc:37]: I missed it in earlier reviews, the overrideEnv section should be of the format:
```
overrideEnv:
  - name: HTTP_PROXY
     value: http://<PROXY_URL>
  - name: HTTPS_PROXY
     value: https://<PROXY_URL>
   - name: NO_PROXY
     value: <IGNORE_PROXY_DOMAINS>
```
Where <proxy_url> is the url to the proxy server and <ignore_proxy_domains> are comma separated list of domains which proxy should ignore eg. “.example.com,*.local.net”.

xref: https://access.redhat.com/documentation/en-us/openshift_container_platform/3.4/html/installation_and_configuration/install-config-http-proxies

**Review by swghosh — APPROVED** [review]
Thanks for addressing earlier comments, seems to be in good shape now.
Just a small change from the overrideEnvs section otherwise mostly lgtm.

**Comment by swghosh** [comment]
@xingxingxia feel free to review the PR at this point for QE perspective!

Note: this skips the part about overrideLabels which we added recently in https://github.com/openshift/cert-manager-operator/pull/112 because we don't yet have a solid use case for the same as overrideLabels validates against `azure.workload.identity/use` which is not useful unless we get complete support for Azure Pod/Workload Identity on our operator.

File: modules/cert-manager-override-environment-variables.adoc
Comment (xingxingxia) [reply modules/cert-manager-override-environment-variables.adoc:37]: Yes, this needs update. https://github.com/openshift/cert-manager-operator/blob/master/deploy/examples/cluster-cert-manager-overrides.yaml shows format example as well

File: modules/cert-manager-override-environment-variables.adoc
Comment (xingxingxia) [inline modules/cert-manager-override-environment-variables.adoc:47]: Pod will be redeployed after configuring.
```suggestion
* Verify that the cert-manager pods are redeployed by entering the following command:
```

File: modules/cert-manager-override-environment-variables.adoc
Comment (xingxingxia) [inline modules/cert-manager-override-environment-variables.adoc:66]: The indentations are not correct. And `POD_NAMESPACE` needs not be documented, because unrelated.
```suggestion
$ oc get pod <redeployed_cert-manager_pod> -n cert-manager -o yaml
----
+
.Example output. The overridden environment variables are added under the `env` field.
[source,yaml]
----
    - name: HTTP_PROXY
      value: http://<PROXY_URL>
    - name: HTTPS_PROXY
      value: https://<PROXY_URL>
    - name: NO_PROXY
      value: <IGNORE_PROXY_DOMAINS>
```

File: modules/cert-manager-override-arguments.adoc
Comment (xingxingxia) [inline modules/cert-manager-override-arguments.adoc:54]:
```suggestion
* Verify that the cert-manager pod is redeployed by entering the following command:
```

Comment (xingxingxia) [inline modules/cert-manager-override-arguments.adoc:54]:
```suggestion
* Verify that the cert-manager pod is redeployed by entering the following command:
```

File: modules/cert-manager-override-arguments.adoc
Comment (xingxingxia) [inline modules/cert-manager-override-arguments.adoc:93]: `oc get -o yaml` will print fields alphabetically.
And the overridden arguments will be shown along with existing ones.
```suggestion
$ oc get pod <redeployed_cert-manager_pod> -n cert-manager -o yaml
----
+
.Example output. The overridden arguments will be added under the `env` field along with existing ones.
[source,yaml]
----
spec:
  containers:
  - args:
    - --acme-http01-solver-nameservers=1.1.1.1:53
    - --cluster-resource-namespace=$(POD_NAMESPACE)
    - --dns01-recursive-nameservers=1.1.1.1:53
    - --dns01-recursive-nameservers-only
    - --leader-election-namespace=kube-system
    - --max-concurrent-challenges=60
    - --metrics-listen-address=<host>:<port>
    - --v=6
```

File: modules/cert-manager-override-arguments.adoc
Comment (xingxingxia) [inline modules/cert-manager-override-arguments.adoc:51]: Maybe you need add a note that, the supported overridden arguments are intentionally limited. Otherwise oc `get pod <redeployed_cert-manager_pod> -n cert-manager -o yaml` does show other arguments which can't be overridden, which can confuse users then.
See https://github.com/openshift/cert-manager-operator/blob/master/pkg/controller/deployment/deployment_overrides_validation.go#L19
@swghosh can have more suggestion if any.

**Comment by xingxingxia** [comment]
Done review from QE side with above comments. No other issues. Thx!

File: modules/cert-manager-override-arguments.adoc
Comment (xingxingxia) [reply modules/cert-manager-override-arguments.adoc:93]: You didn't document `webhookConfig` and `cainjectorConfig` for other two components `cert-manager-cainjector` and `cert-manager-webhook`, so you need not display cert-manager-cainjector-xxxx and cert-manager-webhook-xxxx pods. But when I later review https://github.com/openshift/openshift-docs/pull/59777/files , I see it mentions `To use different log levels for different cert-manager components, see Customizing cert-manager by using the cert-manager Operator API fields`, but `Customizing cert-manager by using the cert-manager Operator API fields` does not document the how-to use different log levels for other two components `cert-manager-cainjector` and `cert-manager-webhook`. So you need to document for them by adding below in `oc edit certmanager cluster` part:
```
  webhookConfig:
    overrideArgs:
      - '--v=6'
  cainjectorConfig:
    overrideArgs:
      - '--v=6'
```

File: modules/cert-manager-override-arguments.adoc
Comment (xenolinux) [inline modules/cert-manager-override-arguments.adoc:9]: @xingxingxia Can we say - "You can override the **supported** arguments...."? Since we cannot override existing i.e default arguments, I thought of using "supported arguments"

Comment (xingxingxia) [reply modules/cert-manager-override-arguments.adoc:9]: OK

File: modules/cert-manager-override-arguments.adoc
Comment (xenolinux) [reply modules/cert-manager-override-arguments.adoc:41]: okay

**Comment by xingxingxia** [comment]
@xenolinux note, this PR and https://github.com/openshift/openshift-docs/pull/59777 features are only implemented for cert-manager Oprator v1.11.1, but PRs' preview does not note this, will this mislead customers if they are using released v1.10.2 while reading the doc after PRs merged?

File: modules/cert-manager-override-arguments.adoc
Comment (swghosh) [reply modules/cert-manager-override-arguments.adoc:42]: We have a fix up for [OCPBUGS-14129](https://issues.redhat.com/browse/OCPBUGS-14129) in openshift/cert-manager-operator#128, (which now seems to be unrelated to overrideArgs field) so I think we can unblock docs for this.

File: modules/cert-manager-override-arguments.adoc
Comment (swghosh) [reply modules/cert-manager-override-arguments.adoc:58]: Sure, we can add a note about the currently supported arg list and that it is currently limited to include only the args that have support with our operator.
Users who wish to use beyond them can use `spec.unsupportedConfigOverrides` (which should be specified with warning/red, etc. as that is well **unsupported**).

**Comment by swghosh** [comment]
> feaatures are only implemented for cert-manager Oprator v1.11.1, ... will this mislead customers if they are using released v1.10.2

@xingxingxia Thanks for bringing this up.
+1, all the doc PRs that we have up for cert-manager operator as of today are specific to v1.11.1 release only and thus we should ensure that these docs be merged right around our actual release date for 1.11. 

@xenolinux please take a look where it'd be worth adding about the release version for the new docs that we're adding, to convey users that these are only supported in 1.11.

The only exception to this was the post-install steps and metrics related doc (#59722, #59703) which was recently merged, which is agreeable because we support them in 1.10 as well.

**Comment by xenolinux** [comment]
> > feaatures are only implemented for cert-manager Oprator v1.11.1, ... will this mislead customers if they are using released v1.10.2
> 
> @xingxingxia Thanks for bringing this up. +1, all the doc PRs that we have up for cert-manager operator as of today are specific to v1.11.1 release only and thus we should ensure that these docs be merged right around our actual release date for 1.11.
> 
> @xenolinux please take a look where it'd be worth adding about the release version for the new docs that we're adding, to convey users that these are only supported in 1.11.
> 
> The only exception to this was the post-install steps and metrics related doc (#59722, #59703) which was recently merged, which is agreeable because we support them in 1.10 as well.

@swghosh Ack. I updated the prerequisite as "The {cert-manager-operator} 1.11.1 is installed." to make it clear that the feature will work on cert-manager operator version 1.11.1. I will update all the PRs. I will be adding release notes as well.

File: modules/cert-manager-override-arguments.adoc
Comment (xenolinux) [reply modules/cert-manager-override-arguments.adoc:58]: @swghosh I added the note as a warning.

File: modules/cert-manager-override-arguments.adoc
Comment (xingxingxia) [inline modules/cert-manager-override-arguments.adoc:14]: Not sure if there will be z-stream v1.11.N. 1.11.1+ is more flexible.
```suggestion
* The {cert-manager-operator} 1.11.1+ is installed.
```

File: modules/cert-manager-override-arguments.adoc
Comment (xingxingxia) [inline modules/cert-manager-override-arguments.adoc:43]: controllerConfig's overrideArgs should follow with controllerConfig immediately
```suggestion
  controllerConfig: <1>
    overrideArgs:
      ...
  webhookConfig:
    overrideArgs:
      - '--v=4'
  cainjectorConfig:
    overrideArgs:
      - '--v=2'
```

File: security/cert_manager_operator/cert-manager-customizing-api-fields.adoc
Comment (xingxingxia) [inline security/cert_manager_operator/cert-manager-customizing-api-fields.adoc:13]: ``
`You cannot override the default arguments` is precisely incorrect. Users can (and sometimes want to) do it per @swghosh https://github.com/openshift/openshift-docs/pull/59726#discussion_r1208897243 .
```suggestion
You can override only limited supported arguments in the {cert-manager-operator}. You cannot override the default arguments.
```

File: modules/cert-manager-override-arguments.adoc
Comment (snarayan-redhat) [inline modules/cert-manager-override-arguments.adoc:50]: I think this callout can be skipped since the step itself mentions that.

File: modules/cert-manager-override-arguments.adoc
Comment (snarayan-redhat) [inline modules/cert-manager-override-arguments.adoc:45]: Are these values supposed to be fixed? If no, then it can be changed to something like callout <5>

File: modules/cert-manager-override-arguments.adoc
Comment (snarayan-redhat) [inline modules/cert-manager-override-arguments.adoc:48]: Are these values supposed to be fixed? If no, then it can be changed to something like callout <5>

File: modules/cert-manager-override-arguments.adoc
Comment (snarayan-redhat) [inline modules/cert-manager-override-arguments.adoc:85]: since this is example output, I think it would be better to have an actual host and port values. Up to you though.

File: modules/cert-manager-override-environment-variables.adoc
Comment (snarayan-redhat) [inline modules/cert-manager-override-environment-variables.adoc:45]: same comment as in previous module

**Review by snarayan-redhat — COMMENTED** [review]
Nice work. Few minor suggestions.

**Pull request merged**
